### PR TITLE
[CM-1191] Observe Zendesk Chat Events

### PR DESCRIPTION
**Pull request by Sathyan-Elangovan**
## Summary
- Observe events -> message, attachment message, agent leave
- Added chat log logic to sync zendesk and kommunicate messages

## Tasks completed:
- When message event triggered, send message to /rest/ws/zendesk/message/send
- When attachment event triggered, send attachment to /rest/ws/zendesk/file/send
- When Agent ends chat from Zendesk, call Resolve conversation API and resolve the conversation.
- When "Restart conversation" is clicked, a new conversation is created to maintain a single chat thread.
- Handled Chat logs -> If the app is closed and opened again, it will sync the messages sent from Zendesk dashboard until session is live

### CODE REVIEW

File: Sources/Utilities/KMZendeskChatHandler.swift [hunk 1]
```diff
@@ -126,19 +177,29 @@ public class KMZendeskChatHandler : NSObject, JWTAuthenticator {
...
-    
     func connectToZendeskSocket() {
         guard let connectionProvider = Chat.connectionProvider else {
             return
         }
+        // Connecting to zeendesk server
         connectionProvider.connect()
         self.observeChatLogs()
+
+         connectionToken = Chat.instance?.connectionProvider.observeConnectionStatus { [self] (connection) in
```
Comment (mukesh-komm) [inline Sources/Utilities/KMZendeskChatHandler.swift:193]: Do we need weak self here?

File: Sources/Utilities/KMZendeskChatHandler.swift [hunk 2]
```diff
@@ -236,13 +293,25 @@ public class KMZendeskChatHandler : NSObject, JWTAuthenticator {
     }
     
     func processChatLogs(logs: [ChatLog]) {
-        guard lastUserMessageCreatedTime != 0 else { return }
-        let filteredArray = logs.filter{$0.createdTimestamp >= TimeInterval(truncating: lastUserMessageCreatedTime)}
-        processAgentMessage(message: filteredArray)
+        let filteredArray = logs.filter{$0.createdTimestamp >= TimeInterval(truncating: lastSyncTime)}
+        for log in filteredArray {
+            if log.type == .message && log.participant == .agent {
```
Comment (mukesh-komm) [inline Sources/Utilities/KMZendeskChatHandler.swift:298]: We can use a switch statement here, with `log.type` and `log.participant` being a tuple.

File: Sources/Utilities/KMZendeskChatHandler.swift (hunk 1 above)
Comment (Sathyan-Elangovan) [reply Sources/Utilities/KMZendeskChatHandler.swift:193]: we don't need here

File: Sources/Utilities/KMZendeskChatHandler.swift [hunk 3]
```diff
@@ -38,33 +57,78 @@ public class KMZendeskChatHandler : NSObject, JWTAuthenticator {
...
     }
     
-    func observeChatLogs() {
-        guard Chat.connectionProvider?.status == .connected else {
-            connectToZendeskSocket()
-            return
+    public func handedOffToAgent(groupId: String) {
+        lastSyncTime = NSDate().timeIntervalSince1970 * 1000 as NSNumber
+        self.groupId = groupId
+        isHandOffHappened = true
```
[the remark this answers is not in the document]
Comment (Sathyan-Elangovan) [reply Sources/Utilities/KMZendeskChatHandler.swift:80]: if we use calculated properly for `isHnadoffHanppened` we need set group id separately. the same for `zenChatIntialized`, we are passing the zendeskaccount key there.

File: Sources/Utilities/KMZendeskChatHandler.swift [hunk 4]
```diff
@@ -38,33 +57,78 @@ public class KMZendeskChatHandler : NSObject, JWTAuthenticator {
     }
     
     public func initiateZendesk(key: String) {
+        if zenChatIntialized {
```
[the remark this answers is not in the document]
Comment (Sathyan-Elangovan) [reply Sources/Utilities/KMZendeskChatHandler.swift]: done

File: Sources/Utilities/KMZendeskChatHandler.swift (hunk 2 above)
Comment (Sathyan-Elangovan) [reply Sources/Utilities/KMZendeskChatHandler.swift]: done

**Review by mukesh-komm — APPROVED** [review]

**Pull request merged**